modules/extension/mbstyle/src/main/java/org/geotools/mbstyle/parse/MBFilter.java

```diff
@@ -460,4 +464,116 @@ public Filter filter() {
             throw new MBFormatException("Unsupported filter " + json);
         }
     }
+
+    /**
+     * Returns true if the input values are equal, false otherwise. The inputs must be numbers,
+     * strings, or booleans, and both of the same type. Examples: ["==", number, number]: boolean
+     * ["==", string, string]: boolean ["==", boolean, boolean]: boolean ["==", null, null]: boolean
+     *
+     * @return equal to expression
+     */
+    private Filter filterEqualTo(JSONArray array) {
+        if (array.size() != 3) {
+            throwUnexpectedArgumentCount("==", 2);
+        }
+        if (parse.isString(array, 1)) { // legacy filter syntax
+            String key = parse.get(array, 1);
+            Object value = parse.value(array, 2);
+            return ff.equal(ff.property(key), ff.literal(value), false);
+        } else {
+            Expression expression1 = parse.string(array, 1);
+            Expression expression2 = parse.string(array, 2);
+            return ff.equal(expression1, expression2, false);
+        }
+    }
+
+    /**
+     * Returns true if the input values are not equal, false otherwise. The inputs must be numbers,
+     * strings, or booleans, and both of the same type. Examples:["!=", number, number]: boolean
+     * ["!=", string, string]: boolean ["!=", boolean, boolean]: boolean ["!=", null, null]: boolean
+     *
+     * @return Not equals expression
+     */
+    private Filter filterNotEqual(JSONArray array) {
+        if (array.size() != 3) {
+            throwUnexpectedArgumentCount("!=", 2);
+        }
+        if (parse.isString(array, 1)) { // legacy filter syntax
+            String key = parse.get(json, 1);
+            Object value = parse.value(json, 2); // legacy filter restricted to literals
+            return ff.notEqual(ff.property(key), ff.literal(value), false);
+        } else {
+            Expression expression1 = parse.string(json, 1);
+            Expression expression2 = parse.string(json, 2);
+            return ff.notEqual(expression1, expression2);
+        }
+    }
+
+    private Filter filterLessOrEqual(JSONArray array) {
+        if (json.size() != 3) {
```

Comment:
looks like there is a lot of copy-and-paste code here and all the other functions.  They're all the same except the last line.

Reply:
Yes, i could probably be smarter on this one, but for now I was trying to exactly match the MBExpression logic. The use of `string` to evaluate expressions is a problem and I would like to have everything consistent before fixing expressions in another pull request.

Still I may be able to factor out a `expression1(json)` and `expression2(json)` method and make this more readable.

Reply:
This has now been resolved, I think the code is easier to understand.